io/src/lib.rs

```diff
@@ -87,3 +87,87 @@ pub mod io {
         }
     }
 }
+
+#[doc(hidden)]
+#[cfg(feature = "core2")]
+/// Re-export core2 for the below macro
+pub use core2 as _core2;
+
+#[doc(hidden)]
+#[cfg(feature = "std")]
+/// Re-export std for the below macro
+pub use std as _std;
+
+#[macro_export]
+/// Because we cannot provide a blanket implementation of [`std::io::Write`] for all implementers
+/// of this crate's `io::Write` trait, we provide this macro instead.
+///
+/// This macro will implement `Write` given a `write` and `flush` fn, either by implementing the
+/// crate's native `io::Write` trait directly, or a more generic trait from `std` or `core2` for
+/// users using those features. In any case, this crate's `io::Write` feature will be implemented
+/// for the given type, even if indirectly.
+#[cfg(not(any(feature = "core2", feature = "std")))]
+macro_rules! impl_write {
```

Comment:
In dd8dd88d50d152b627e1ee869bb2cd0360788b4d:

Fine to wait for a followup PR, but all these macros should have invocations that "look like" the code that they generate. Rather than being lists of exprs.